### PR TITLE
Component to angle transform

### PR DESCRIPTION
**Pull request by CharmaineWONG2**
## Summary by CodeRabbit

- **New Features**
  - Enhanced spin transformation functionality now supports both forward and inverse mappings.
  - Added utilities for vector rotations and precise orbital angular momentum computation.
  - Introduced a new script for generating test data related to spin transformations.

- **Refactor**
  - Updated parameter naming and improved conditional configuration for a more intuitive setup.

- **Tests**
  - Expanded the testing suite to rigorously validate transformation accuracy and reliability.

### CODE REVIEW

**Review by thomasckng — CHANGES_REQUESTED** [review]

**Comment by SSL32081** [comment]
I have a look at the failed test, it was because the starting location of the tests is at the top directory, so it can’t find the files. 

It can be solved by prepending the file path with `test/unit/`, _e.g._ `test/unit/source_files/spin_angles_input.npz`

File: test/unit/test_transform.py [hunk 1]
```diff
@@ -210,13 +212,44 @@ def test_backward_spin_transform(self):
...
+        q = eta_to_q(eta)
+
+        inputs = jnp.array([iota, S1x, S1y, S1z, S2x, S2y, S2z, M_c, q, fRef, phiRef]).T
+
+        for i in range(100):
+            jimgw_spins = cartesian_spin_to_spin_angles(*inputs[i])
+            jimgw_spins = jnp.concatenate([jnp.array(jimgw_spins), inputs[i][-4:]])
+            jimgw_spins = spin_angles_to_cartesian_spin(*jimgw_spins)
+
+            assert jnp.allclose(jnp.array(jimgw_spins), inputs[i][:7], atol=1e-4)
```
[the remark this answers is not in the document]
Comment (thomasckng) [reply test/unit/test_transform.py:252]: That's the best I could get. @CharmaineWONG2 Can you take a look on this?

[the remark this answers is not in the document]
Comment (thomasckng) [reply test/unit/test_transform.py:252]: Changed the `atol` to `1e-16` for the backward transform. Thanks

File: test/unit/test_transform.py
Comment (SSL32081) [inline test/unit/test_transform.py:1]: Regarding the list of samples, it might be worth checking whether the spin magnitudes are less than unity before appending them to the list of parameters. 

In principle, the maths in the transform itself works even if the spin magnitudes are beyond that, but it sounds better to keep only physically feasible parameters. 

I believe this should amount to adding a few lines:
```python
inputs = []
accepted = 0
        while accepted <= 100:
            iota = np.array(np.random.uniform(0, np.pi))
            S1x = np.array(np.random.uniform(-1, 1))
        ...
            q = symmetric_mass_ratio_to_mass_ratio(eta)
            m1, m2 = chirp_mass_and_mass_ratio_to_component_masses(M_c, q)
            
            a1 = jnp.linalg.norm([S1x, S1y, S1z])
            a2 = jnp.linalg.norm([S2x, S2y, S2z])
            
            if not (a1 <= 1) * (a2 <= 1):
                continue
            inputs.append((iota, S1x, S1y, S1z, S2x, S2y, S2z, m1, m2, fRef, phiRef))
            accepted += 1
```

**Comment by thomasckng** [comment]
@SSL32081 One more issue that I just think of. We use `phase_c` while the bilby function is using `phase_ref`.

**Review by thomasckng — APPROVED** [review]

**Comment by thomasckng** [comment]
@kazewong Added tests for JIT. Ready to merge.

File: test/unit/test_transform.py [hunk 2]
```diff
@@ -1,48 +1,477 @@
...
+            "s1_z",
+            "s2_x",
+            "s2_y",
+            "s2_z",
+            "M_c",
+            "q",
+            "phase_c",
+        )
+
+        ##############################
```
Comment (kazewong) [inline test/unit/test_transform.py:34]: Are these commented out block still needed? If not please take them out

Comment (thomasckng) [reply test/unit/test_transform.py:34]: These are for the generation of the input and output binary files. If you prefer not to keep it, I can delete them now. I just want to make sure the generation code of the binary files are documented somewhere, do you have any suggestions regarding this?

**Comment by thomasckng** [comment]
This PR adds the transform class for transforming between spin angles (theta_jn, a_1, a_2, phi_12,...) and Cartesian spins (s1_x, s1_y,...), based on the previous implementation. The new class includes transformation in both directions.

A unit test was added together in this PR, which includes testing for JIT compilation. The outcome of the transform is tested in both directions by cross checking with bilby conversion functions. The input and output of the bilby functions are stored in binary files. The generation code of the binary files are the following
```
import numpy as np
from lalsimulation import SimInspiralTransformPrecessingNewInitialConditions
from bilby.gw.conversion import (
    symmetric_mass_ratio_to_mass_ratio,
    chirp_mass_and_mass_ratio_to_component_masses,
)
from lal import MSUN_SI

inputs = []
for _ in range(100):
    thetaJN = np.array(np.random.uniform(0, np.pi))
    phiJL = np.array(np.random.uniform(0, 2 * np.pi))
    theta1 = np.array(np.random.uniform(0, np.pi))
    theta2 = np.array(np.random.uniform(0, np.pi))
    phi12 = np.array(np.random.uniform(0, 2 * np.pi))
    chi1 = np.array(np.random.uniform(0, 1))
    chi2 = np.array(np.random.uniform(0, 1))
    M_c = np.array(np.random.uniform(1, 100))
    eta = np.array(np.random.uniform(0.1, 0.25))
    fRef = np.array(np.random.uniform(10, 1000))
    phiRef = np.array(np.random.uniform(0, 2 * np.pi))

    q = symmetric_mass_ratio_to_mass_ratio(eta)
    m1, m2 = chirp_mass_and_mass_ratio_to_component_masses(M_c, q)

    inputs.append(
        (
            thetaJN,
            phiJL,
            theta1,
            theta2,
            phi12,
            chi1,
            chi2,
            m1,
            m2,
            fRef,
            phiRef,
        )
    )
inputs = np.array(inputs)
np.savez(
    "test/unit/source_files/spin_angles_input.npz",
    thetaJN=inputs[:, 0],
    phiJL=inputs[:, 1],
    theta1=inputs[:, 2],
    theta2=inputs[:, 3],
    phi12=inputs[:, 4],
    chi1=inputs[:, 5],
    chi2=inputs[:, 6],
    m1=inputs[:, 7],
    m2=inputs[:, 8],
    fRef=inputs[:, 9],
    phiRef=inputs[:, 10],
)

bilby_outputs = []
for input in inputs:
    iota, S1x, S1y, S1z, S2x, S2y, S2z = (
        SimInspiralTransformPrecessingNewInitialConditions(
            input[0],
            input[1],
            input[2],
            input[3],
            input[4],
            input[5],
            input[6],
            input[7] * MSUN_SI,
            input[8] * MSUN_SI,
            input[9],
            input[10],
        )
    )
    bilby_outputs.append((iota, S1x, S1y, S1z, S2x, S2y, S2z))
bilby_outputs = np.array(bilby_outputs)
np.savez(
    "test/unit/source_files/cartesian_spins_output_for_bilby.npz",
    iota=bilby_outputs[:, 0],
    S1x=bilby_outputs[:, 1],
    S1y=bilby_outputs[:, 2],
    S1z=bilby_outputs[:, 3],
    S2x=bilby_outputs[:, 4],
    S2y=bilby_outputs[:, 5],
    S2z=bilby_outputs[:, 6],
)


import numpy as np
from lalsimulation import SimInspiralTransformPrecessingWvf2PE
from bilby.gw.conversion import (
    symmetric_mass_ratio_to_mass_ratio,
    chirp_mass_and_mass_ratio_to_component_masses,
)

inputs = []
while len(inputs) < 100:
    iota = np.array(np.random.uniform(0, np.pi))
    S1x = np.array(np.random.uniform(-1, 1))
    S1y = np.array(np.random.uniform(-1, 1))
    S1z = np.array(np.random.uniform(-1, 1))
    S2x = np.array(np.random.uniform(-1, 1))
    S2y = np.array(np.random.uniform(-1, 1))
    S2z = np.array(np.random.uniform(-1, 1))
    if (
        jnp.linalg.norm(jnp.array([S1x, S1y, S1z])) >= 1
        or jnp.linalg.norm(jnp.array([S2x, S2y, S2z])) >= 1
    ):
        continue
    M_c = np.array(np.random.uniform(1, 100))
    eta = np.array(np.random.uniform(0.1, 0.25))
    fRef = np.array(np.random.uniform(10, 100))
    phiRef = np.array(np.random.uniform(0, 2 * np.pi))

    q = symmetric_mass_ratio_to_mass_ratio(eta)
    m1, m2 = chirp_mass_and_mass_ratio_to_component_masses(M_c, q)

    inputs.append((iota, S1x, S1y, S1z, S2x, S2y, S2z, m1, m2, fRef, phiRef))
inputs = np.array(inputs)
np.savez(
    "test/unit/source_files/cartesian_spins_input.npz",
    iota=inputs[:, 0],
    S1x=inputs[:, 1],
    S1y=inputs[:, 2],
    S1z=inputs[:, 3],
    S2x=inputs[:, 4],
    S2y=inputs[:, 5],
    S2z=inputs[:, 6],
    m1=inputs[:, 7],
    m2=inputs[:, 8],
    fRef=inputs[:, 9],
    phiRef=inputs[:, 10],
)

bilby_outputs = []
for input in inputs:
    thteaJN, phiJL, theta1, theta2, phi12, chi1, chi2 = (
        SimInspiralTransformPrecessingWvf2PE(*input)
    )
    bilby_outputs.append((thteaJN, phiJL, theta1, theta2, phi12, chi1, chi2))
bilby_outputs = np.array(bilby_outputs)
np.savez(
    "test/unit/source_files/spin_angles_output_for_bilby.npz",
    thetaJN=bilby_outputs[:, 0],
    phiJL=bilby_outputs[:, 1],
    theta1=bilby_outputs[:, 2],
    theta2=bilby_outputs[:, 3],
    phi12=bilby_outputs[:, 4],
    chi1=bilby_outputs[:, 5],
    chi2=bilby_outputs[:, 6],
)
```

**Comment by thomasckng** [comment]
@coderabbitai review

File: test/unit/test_transform.py [hunk 3]
```diff
@@ -1,48 +1,291 @@
...
+        subkeys = jax.random.split(subkey, 11)
+        iota = jax.random.uniform(subkeys[0], (1,), minval=0, maxval=jnp.pi)
+        while True:
+            S1x = jax.random.uniform(subkeys[1], (1,), minval=-1, maxval=1)
+            S1y = jax.random.uniform(subkeys[2], (1,), minval=-1, maxval=1)
+            S1z = jax.random.uniform(subkeys[3], (1,), minval=-1, maxval=1)
+            S2x = jax.random.uniform(subkeys[4], (1,), minval=-1, maxval=1)
+            S2y = jax.random.uniform(subkeys[5], (1,), minval=-1, maxval=1)
+            S2z = jax.random.uniform(subkeys[6], (1,), minval=-1, maxval=1)
+            if jnp.linalg.norm(jnp.array([S1x[0], S1y[0], S1z[0]])) >= 1:
```
Comment (thomasckng) [inline test/unit/test_transform.py:181]: Minor bug here. @CharmaineWONG2 is fixing it.

**Comment by thomasckng** [comment]
@coderabbitai review

**Comment by thomasckng** [comment]
@kazewong I will sync `jim-dev` to this branch and solve the conflicts tomorrow. Maybe you can take a look after that and merge this before migrating to the new `flowMC` version. This transform is tested, but we are not using it in any of our runs currently.

**Comment by kazewong** [comment]
There are still standing comments that have not been addressed. Please address them before approving the changes

**Review by kazewong — COMMENTED** [review]
Add review

File: test/unit/test_transform.py [hunk 4]
```diff
@@ -1,48 +1,477 @@
...
+        #     chi2=bilby_outputs[:, 6],
+        # )
+
+        # read inputs from binary
+        inputs = jnp.load("test/unit/source_files/cartesian_spins_input.npz")
+        inputs = [inputs[key] for key in inputs.keys()]
+        M_c, q = m1_m2_to_Mc_q(inputs[7], inputs[8])
+
+        # compute jimgw spins
+        for i in range(100):
```
[the remark this answers is not in the document]
Comment (thomasckng) [reply test/unit/test_transform.py]: The reason for not using `vmap` is that the transform class has to be redefined for every sample, as each sample has a different `f_ref`. However, I agree we should modify this a bit.

@CharmaineWONG2 How about we draw random samples for a few different `f_ref` instead?

File: src/jimgw/single_event/utils.py [hunk 5]
```diff
@@ -456,87 +495,67 @@ def spin_to_cartesian_spin(
...
+    s2 = m2 * m2 * chi_2 * s2hat
     J = s1 + s2 + jnp.array([0.0, 0.0, Lmag])
 
+    # Normalize J, and find theta0 and phi0 (the angles in starting frame)
     Jhat = J / jnp.linalg.norm(J)
     theta0 = jnp.arccos(Jhat[2])
     phi0 = jnp.arctan2(Jhat[1], Jhat[0])
 
-    # Rotation 1:
+    # Rotation 1: Rotate about z-axis by -phi0
```
[the remark this answers is not in the document]
Comment (SSL32081) [reply src/jimgw/single_event/utils.py:619]: It appears that the Euler rotation functions are quite different from the `rotation_y,z`, which might introduce extra inaccuracies, and some additional checks might also be needed for corner cases to ensure the mapping is exactly the same. 

Since we have verified the correctness and accuracy of this transform (and the rotations), it would be better to leave this as a future optimisation issue.

File: src/jimgw/single_event/utils.py [hunk 6]
```diff
@@ -480,14 +480,89 @@ def ra_dec_to_zenith_azimuth(
...
+        Relative velocity at the reference frequency.
+    Returns
+    -------
+    Lmag : Float
+        Magnitude of the orbital angular momentum.
+    """
+    eta = m1 * m2 / (m1 + m2) ** 2
+    ## Simplified from:
+    ## (m1 + m2) * (m1 + m2) * eta = m1 * m2
+    LN = (m1 + m2) * (m1 + m2) * eta / v0
```
Comment (SSL32081) [inline src/jimgw/single_event/utils.py:553]: I think this comment was here because it was simplified to just
```
LN = m1 * m2 / v0 
```
If we are keeping the long form, I think the comment can be removed.

File: test/unit/source_files/bilby_spin.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    inputs = []
+    for _ in range(100):
+        iota = np.array(np.random.uniform(0, np.pi))
+        while True:
+            S1x = np.array(np.random.uniform(-1, 1))
+            S1y = np.array(np.random.uniform(-1, 1))
+            S1z = np.array(np.random.uniform(-1, 1))
+            S2x = np.array(np.random.uniform(-1, 1))
+            S2y = np.array(np.random.uniform(-1, 1))
+            S2z = np.array(np.random.uniform(-1, 1))
```
Comment (SSL32081) [inline test/unit/source_files/bilby_spin.py:111]: The additional `np.array` wrapping seems unnecessary.

File: test/unit/source_files/bilby_spin.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+            if (
+                jnp.linalg.norm(jnp.array([S1x, S1y, S1z])) <= 1
+                or jnp.linalg.norm(jnp.array([S2x, S2y, S2z])) <= 1
+            ):
+                break
+            M_c = np.array(np.random.uniform(1, 100))
+            eta = np.array(np.random.uniform(0.1, 0.25))
+            phiRef = np.array(np.random.uniform(0, 2 * np.pi))
```
Comment (SSL32081) [inline test/unit/source_files/bilby_spin.py:119]: The additional `np.array` wrapping seems unnecessary.

**Comment by thomasckng** [comment]
@coderabbitai summary

File: test/unit/source_files/bilby_spin.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    np.savez(
+        f"test/unit/source_files/cartesian_spins_output_for_bilby/cartesian_spins_output_for_bilby_fRef_{fRef}.npz",
+        iota=bilby_outputs[:, 0],
+        S1x=bilby_outputs[:, 1],
+        S1y=bilby_outputs[:, 2],
+        S1z=bilby_outputs[:, 3],
+        S2x=bilby_outputs[:, 4],
+        S2y=bilby_outputs[:, 5],
+        S2z=bilby_outputs[:, 6],
+    )
```
Comment (SSL32081) [inline test/unit/source_files/bilby_spin.py:94]: The entire loop can be simplified to:
```python
for fRef in [10, 30, 50]:
    N_samples = 100

    M_c = np.random.uniform(1, 100)
    eta = np.random.uniform(0.1, 0.25)
    q = symmetric_mass_ratio_to_mass_ratio(eta)
    m1, m2 = chirp_mass_and_mass_ratio_to_component_masses(M_c, q)

    input_dict = {
        'thetaJN': np.random.uniform(0, np.pi, N_samples),
        'phiJL': np.random.uniform(0, 2 * np.pi),
        'theta1': np.random.uniform(0, np.pi),
        'theta2': np.random.uniform(0, np.pi),
        'phi12': np.random.uniform(0, 2 * np.pi),
        'chi1': np.random.uniform(0, 1),
        'chi2': np.random.uniform(0, 1),
        'm1': m1,
        'm2': m2,
        'fRef': np.ones(100) * fRef,
        'phiRef': np.random.uniform(0, 2 * np.pi),
    }

    np.savez(
        f"test/unit/source_files/spin_angles_input/spin_angles_input_fRef_{fRef}.npz",
        **input_dict
    )

    input_dict['m1'] *= MSUN_SI
    input_dict['m2'] *= MSUN_SI

    input_array = np.array(list(input_dict.values())).T

    bilby_outputs = []
    for row in input_array:
        bilby_outputs.append(
            SimInspiralTransformPrecessingNewInitialConditions(*row))

    bilby_outputs = np.array(bilby_outputs).T

    np.savez(
        f"test/unit/source_files/cartesian_spins_output_for_bilby/cartesian_spins_output_for_bilby_fRef_{fRef}.npz",
        iota=bilby_outputs[0],
        S1x = bilby_outputs[1],
        S1y = bilby_outputs[2],
        S1z = bilby_outputs[3],
        S2x = bilby_outputs[4],
        S2y = bilby_outputs[5],
        S2z = bilby_outputs[6]
    )
```

**Comment by thomasckng** [comment]
@coderabbitai summary

**Comment by thomasckng** [comment]
@coderabbitai summary

**Comment by thomasckng** [comment]
@kazewong ready for second review.

**Comment by thomasckng** [comment]
@SSL32081 Thanks for resolving the conflicts. @kazewong It's ready now.
